File: scala-source/com/barrybecker4/puzzle/redpuzzle/solver/RedPuzzleSolver.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.redpuzzle.solver

import com.barrybecker4.math.MathUtil
import com.barrybecker4.puzzle.common.PuzzleController
import com.barrybecker4.puzzle.common.solver.PuzzleSolver
import com.barrybecker4.puzzle.redpuzzle.model.{OrientedPiece, PieceList, PieceLists}


/**
  * Abstract base class for puzzle solver strategies (see strategy pattern).
  * Subclasses do the hard work of actually solving the puzzle.
  * @param puzzle the puzzle to solve.
  * @param pieces initial pool of shuffled pieces to place (defaults to a random standard set).
  * @author Barry Becker
  */
abstract class RedPuzzleSolver(
  val puzzle: PuzzleController[PieceList, OrientedPiece],
  val pieces: PieceList = PieceLists.getInitialPuzzlePieces(MathUtil.RANDOM)
) extends PuzzleSolver[OrientedPiece] {

  /** the pieces we have correctly fitted so far. */
  var solution: PieceList = PieceList(List.empty[OrientedPiece], pieces.numTotal)

  /** some measure of the number of iterations the solver needs to solve the puzzle. */
  var numTries: Int = 0

  /** Solve the puzzle
    * @return true if a solution is found.
    */
  def solve: Option[Seq[OrientedPiece]]
}

File: source/com/barrybecker4/puzzle/redpuzzle1/solver/RedPuzzleSolver.java
// Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.redpuzzle1.solver;

import com.barrybecker4.puzzle.common1.PuzzleController;
import com.barrybecker4.puzzle.common1.solver.PuzzleSolver;
import com.barrybecker4.puzzle.redpuzzle1.model.Piece;
import com.barrybecker4.puzzle.redpuzzle1.model.PieceList;
import com.barrybecker4.puzzle.redpuzzle1.model.PieceLists;
import scala.Option;
import scala.collection.Seq;


/**
 * Abstract base class for puzzle solver strategies (see strategy pattern).
 * Subclasses do the hard work of actually solving the puzzle.
 * Controller in the model-view-controller pattern.
 *
 * @author Barry Becker
 */
public abstract class RedPuzzleSolver
                implements PuzzleSolver<Piece> {

    /** the unsorted pieces that we draw from and place in the solvedPieces list. */
    PieceList pieces_;

    /** the pieces we have correctly fitted so far. */
    PieceList solution_;

    /** some measure of the number of iterations the solver needs to solve the puzzle. */
    int numTries_ = 0;

    PuzzleController<PieceList, Piece> puzzle;

    /**
     * Constructor
     * @param puzzle the puzzle to solve.
     */
    RedPuzzleSolver(PuzzleController<PieceList, Piece> puzzle) {
        this.puzzle = puzzle;
        pieces_ = PieceLists.getInitialPuzzlePieces();
        solution_ = new PieceList();
    }

    /**
     * Derived classes must provide the implementation for this abstract method.
     * @return true if a solution is found.
     */
    @Override
    public abstract Option<Seq<Piece>> solve();

}
